### PR TITLE
Corrected Strict NON NULL return type checks

### PR DESCRIPTION
**Pull request by sreemanamala**
### Description

aims to fix strict type compare issue (Boolean vs Boolean not null) in TIME_IN_INTERVAL

<hr>

##### Key changed/added classes in this PR
 * `TimeInIntervalConvertletFactory.java`
 * `ContainsOperatorConversion.java`
 * `RegexpLikeOperatorConversion.java`
 * `SubstringOperatorConversion.java`

<hr>

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/convertlet/TimeInIntervalConvertletFactory.java [hunk 1]
```diff
@@ -56,7 +56,7 @@ public class TimeInIntervalConvertletFactory implements DruidConvertletFactory
       .operandTypes(SqlTypeFamily.TIMESTAMP, SqlTypeFamily.CHARACTER)
       .requiredOperandCount(2)
       .literalOperands(1)
-      .returnTypeNonNull(SqlTypeName.BOOLEAN)
+      .returnTypeInference(ReturnTypes.BOOLEAN_NULLABLE)
```
Comment (kgyrtkirk) [inline sql/src/main/java/org/apache/druid/sql/calcite/planner/convertlet/TimeInIntervalConvertletFactory.java:59]: could you take a look at other methods not permitting `null`-on outputs?

it seems like that for example `regexp_like` is also affected:
```
        SELECT REGEXP_LIKE(dim1,'x') from foo where dim1 is null
```

Comment (sreemanamala) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/convertlet/TimeInIntervalConvertletFactory.java:59]: fixed in regexp_like, substring & contains operators. rest looked good to me.

File: processing/src/main/java/org/apache/druid/math/expr/Function.java [hunk 2]
```diff
@@ -3634,7 +3634,7 @@ ExprEval doApply(ExprEval arrayExpr, ExprEval scalarExpr)
       final Object[] array = arrayExpr.asArray();
       final int position = scalarExpr.asInt() - 1;
 
-      if (array.length > position) {
+      if (array.length > position && position >= 0) {
```
Comment (kgyrtkirk) [inline processing/src/main/java/org/apache/druid/math/expr/Function.java:3637]: is this change was coverd with a test? if this is a real issue - I've noticed the same mistake in `ArrayOffsetFunction` :)

File: sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/SubstringOperatorConversion.java [hunk 3]
```diff
@@ -45,7 +46,7 @@ public class SubstringOperatorConversion implements SqlOperatorConversion
       .operatorBuilder("SUBSTRING")
       .operandTypes(SqlTypeFamily.CHARACTER, SqlTypeFamily.INTEGER, SqlTypeFamily.INTEGER)
       .functionCategory(SqlFunctionCategory.STRING)
-      .returnTypeInference(ReturnTypes.ARG0)
+      .returnTypeInference(ReturnTypes.ARG0.andThen(SqlTypeTransforms.FORCE_NULLABLE))
```
Comment (kgyrtkirk) [inline sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/SubstringOperatorConversion.java:49]: afaik these are appied during calcite planning ; I see some tests in `ExpressionTest` ; aren't those only excercise the native layer?

File: sql/src/test/java/org/apache/druid/sql/calcite/expression/ExpressionsTest.java [hunk 4]
```diff
@@ -270,6 +270,16 @@ public void testSubstring()
         makeExpression("substring(\"spacey\", (\"p\" - 1), \"p\")"),
         "hey"
     );
+    testHelper.testExpressionString(
```
Comment (kgyrtkirk) [inline sql/src/test/java/org/apache/druid/sql/calcite/expression/ExpressionsTest.java:273]: I've unpatched all non-test changes - and these cases still pass...did I miss something?

File: processing/src/main/java/org/apache/druid/math/expr/Function.java (hunk 2 above)
Comment (sreemanamala) [reply processing/src/main/java/org/apache/druid/math/expr/Function.java:3637]: I didnt add a specific test for this. saw this as an issue & thought of correcting off. Will change in ArrayOffsetFunction as well

File: sql/src/test/java/org/apache/druid/sql/calcite/expression/ExpressionsTest.java (hunk 4 above)
Comment (sreemanamala) [reply sql/src/test/java/org/apache/druid/sql/calcite/expression/ExpressionsTest.java:273]: I guess these tests here are only running on the native layer. And the issue which we face is more of coming in the calcite layer. Let me change the tests. Thanks for pointing these.

Moreover 
```select CONTAINS_STRING(null, 'a') -- gives INVALID use of null
select CONTAINS_STRING(dim3, 'a') from numFoo where dim3 is null -- fails the strict check```

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java [hunk 5]
```diff
@@ -15612,4 +15642,51 @@ public void testBitwiseXor()
...
+                                            "v1",
+                                            "null",
+                                            ColumnType.STRING,
+                                            ExprMacroTable.nil()
+                                    )
+                            )
+                            .columns("v0", "v1")
+                            .filters(
+                                    NullHandling.replaceWithDefault()
+                                            ? selector("dim3", null)
```
Comment (kgyrtkirk) [inline sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java:15674]: `isnull("dim")` has `replaceWithDefault` inside - so no need outside

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java [hunk 6]
```diff
@@ -15612,4 +15642,51 @@ public void testBitwiseXor()
         )
     );
   }
+
+  @Test
+  public void testStringOperationsBooleanNullable()
+  {
+    testQuery(
+            "SELECT ICONTAINS_STRING(dim3, 'a'), REGEXP_LIKE(dim3,'x'), SUBSTRING(dim3, 1, 1) " +
+                    "from druid.numfoo where dim3 is NULL",
```
Comment (kgyrtkirk) [inline sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java:15651]: add `limit 1` => no more 2/3 results

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java [hunk 7]
```diff
@@ -15612,4 +15642,51 @@ public void testBitwiseXor()
...
+                            .build()
+            ),
+            NullHandling.replaceWithDefault() ? ImmutableList.of(
+                    new Object[]{false, false, ""},
+                    new Object[]{false, false, ""},
+                    new Object[]{false, false, ""}
+            ) : ImmutableList.of(
+                    new Object[]{null, null, null},
+                    new Object[]{null, null, null}
+            )
```
Comment (kgyrtkirk) [inline sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java:15689]: `testBuilder` can be switched to compare in `RELAX_NULLS` mode

**Review by kgyrtkirk — APPROVED** [review]
some minor test related notes

**Pull request merged**